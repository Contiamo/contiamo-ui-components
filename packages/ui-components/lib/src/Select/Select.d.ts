export declare type Value = number | string;
export interface Option {
    label?: string;
    value: Value;
}
export interface IProps {
    id?: string;
    domId?: string;
    css?: any;
    className?: string;
    options: Option[];
    value: undefined | Value | Value[];
    filterable?: boolean;
    disabled?: boolean;
    onChange: (newValue: Value | Value[], changedItem?: Value) => void;
    onClick?: () => void;
    onFilter?: () => void;
    color?: string;
    placeholder?: string;
}
declare const _default: any;
export default _default;
