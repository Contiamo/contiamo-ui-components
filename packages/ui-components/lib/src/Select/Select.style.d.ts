/// <reference types="react" />
import { GlamorousComponent } from "glamorous";
import { Theme } from "contiamo-ui-theme";
declare const Container: GlamorousComponent<Pick<{
    theme: Theme;
    color?: string;
    disabled: boolean;
    updating: boolean;
    style?: any;
}, "style" | "color" | "disabled" | "updating"> & React.HTMLProps<HTMLDivElement>, {
    theme: Theme;
    color?: string;
    disabled: boolean;
    updating: boolean;
    style?: any;
}>;
declare const DisplayValue: GlamorousComponent<Pick<{
    theme: Theme;
    isPlaceholder: boolean;
}, "isPlaceholder"> & React.HTMLProps<HTMLDivElement>, {
    theme: Theme;
    isPlaceholder: boolean;
}>;
declare const Options: GlamorousComponent<Pick<{
    theme: Theme;
}, never> & React.HTMLProps<HTMLDivElement>, {
    theme: Theme;
}>;
declare const OptionsList: GlamorousComponent<Pick<{
    theme: Theme;
}, never> & React.HTMLProps<HTMLDivElement>, {
    theme: Theme;
}>;
export { Container, Options, OptionsList, DisplayValue };
