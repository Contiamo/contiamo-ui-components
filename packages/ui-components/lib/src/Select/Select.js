"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var glamorous_1 = require("glamorous");
var SelectOption_1 = require("./Option/SelectOption");
var SelectFilter_1 = require("./Filter/SelectFilter");
var Select_style_1 = require("./Select.style");
var Select = /** @class */ (function (_super) {
    __extends(Select, _super);
    function Select(props) {
        var _this = _super.call(this, props) || this;
        // This implements "click outside to close" behavior
        _this.handleClick = function (e) {
            var el = _this.container;
            // if we're somehow not working with a DOM node (flowtype is fun!)
            if (!(e.target instanceof Node)) {
                return;
            }
            // if we're clicking on the Select itself,
            if (el && el.contains(e.target)) {
                return;
            }
            // if we're clicking outside,
            _this.close();
        };
        _this.handleEsc = function (e) {
            if (e.keyCode === 27) {
                _this.close();
            }
        };
        _this.state = {
            open: false,
            updating: false,
            filter: new RegExp(/./)
        };
        return _this;
    }
    Select.prototype.componentDidMount = function () {
        window.addEventListener("click", this.handleClick, true);
        window.addEventListener("keyup", this.handleEsc, true);
    };
    Select.prototype.componentWillUnmount = function () {
        window.removeEventListener("click", this.handleClick, true);
        window.removeEventListener("keyup", this.handleEsc, true);
    };
    Select.prototype.getDisplayValue = function () {
        if (!this.props.value) {
            return typeof this.props.placeholder === "string" ? this.props.placeholder : "";
        }
        if (!Array.isArray(this.props.value)) {
            return this.props.value.label;
        }
        return this.props.value.map(function (option) { return option.label; }).slice().join(", ");
    };
    Select.prototype.selectOption = function (option) {
        if (!Array.isArray(this.props.value)) {
            this.props.onChange(option);
            return;
        }
        var optionIndex = this.props.value.indexOf(option);
        if (optionIndex < 0) {
            this.props.onChange(this.props.value.concat([option]).filter(function (item) { return !item.placeholder; }));
        }
        else {
            this.props.onChange(this.props.value.slice(0, optionIndex).concat(this.props.value.slice(optionIndex + 1)));
        }
    };
    Select.prototype.isOptionSelected = function (option) {
        if (!Array.isArray(this.props.value)) {
            return this.props.value === option;
        }
        return this.props.value.indexOf(option) > -1;
    };
    Select.prototype.updateFilter = function (event) {
        return __awaiter(this, void 0, void 0, function () {
            var filter;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        event.persist();
                        if (!(event.target instanceof HTMLInputElement)) {
                            throw new Error("<Select>: Your filter field is _not_ an input element and therefore has an unreadable value.");
                        }
                        if (!this.props.onFilter) return [3 /*break*/, 2];
                        this.setState(function () { return ({ updating: true }); });
                        return [4 /*yield*/, this.props.onFilter()];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2:
                        filter = new RegExp(event.target.value, "i");
                        this.setState(function () { return ({ filter: filter, updating: false }); });
                        return [2 /*return*/];
                }
            });
        });
    };
    Select.prototype.toggle = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(!this.state.open && this.props.onClick)) return [3 /*break*/, 2];
                        this.setState(function () { return ({ updating: true }); });
                        return [4 /*yield*/, this.props.onClick()];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2:
                        this.setState(function (prevState) { return ({ updating: false, open: !prevState.open }); });
                        return [2 /*return*/];
                }
            });
        });
    };
    Select.prototype.close = function () {
        this.setState(function () { return ({
            open: false
        }); });
    };
    Select.prototype.render = function () {
        var _this = this;
        return (React.createElement("div", { ref: function (container) { return (_this.container = container); }, className: this.props.className + " Select" + (this.state.updating ? " Select_updating" : ""), role: "listbox", tabIndex: -2, onClick: function () { return _this.toggle(); } },
            React.createElement("div", { className: "Select__value" }, this.getDisplayValue() || this.props.placeholder),
            this.props.options.length && this.state.open ? (React.createElement("div", { className: "Select__options" },
                this.props.filterable && React.createElement(SelectFilter_1.default, { onChange: function (e) { return _this.updateFilter(e); } }),
                React.createElement("div", { className: "Select__options_list" }, this.props.options.map(function (option) {
                    return option.label.match(_this.state.filter) && (React.createElement(SelectOption_1.default, { key: option.id, onClick: function () { return _this.selectOption(option); }, selected: _this.isOptionSelected(option) }, option.label));
                })))) : ("")));
    };
    Select.defaultProps = {
        className: "",
        filterable: false,
        disabled: false,
        multiple: false
    };
    return Select;
}(React.Component));
exports.Select = Select;
exports.default = glamorous_1.default(Select)(Select_style_1.default);
//# sourceMappingURL=Select.js.map