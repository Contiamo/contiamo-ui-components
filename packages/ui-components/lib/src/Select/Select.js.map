{"version":3,"file":"Select.js","sourceRoot":"","sources":["../../../src/Select/Select.tsx"],"names":[],"mappings":";;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;AAAA,6BAA8B;AAG9B,sDAAgD;AAChD,sDAAgD;AAEhD,+CAAgE;AA6BhE;IAAqB,0BAA+B;IAApD;QAAA,qEAkJC;QAjJC,WAAK,GAAW;YACd,IAAI,EAAE,KAAK;YACX,QAAQ,EAAE,KAAK;YACf,MAAM,EAAE,IAAI,MAAM,CAAC,GAAG,CAAC;SACxB,CAAA;QAKD,oDAAoD;QACpD,iBAAW,GAAG,UAAC,CAAa;YAC1B,IAAM,EAAE,GAAG,KAAI,CAAC,SAAS,CAAA;YAEzB,kEAAkE;YAClE,EAAE,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,MAAM,YAAY,IAAI,CAAC,CAAC,CAAC,CAAC;gBAChC,MAAM,CAAA;YACR,CAAC;YACD,0CAA0C;YAC1C,EAAE,CAAC,CAAC,EAAE,IAAI,EAAE,CAAC,QAAQ,CAAC,CAAC,CAAC,MAAM,CAAC,CAAC,CAAC,CAAC;gBAChC,MAAM,CAAA;YACR,CAAC;YAED,6BAA6B;YAC7B,KAAI,CAAC,KAAK,EAAE,CAAA;QACd,CAAC,CAAA;QAED,eAAS,GAAG,UAAC,CAAgB;YAC3B,EAAE,CAAC,CAAC,CAAC,CAAC,OAAO,KAAK,EAAE,CAAC,CAAC,CAAC;gBACrB,KAAI,CAAC,KAAK,EAAE,CAAA;YACd,CAAC;QACH,CAAC,CAAA;;IAmHH,CAAC;IAjHC,kCAAiB,GAAjB;QACE,MAAM,CAAC,gBAAgB,CAAC,OAAO,EAAE,IAAI,CAAC,WAAW,EAAE,IAAI,CAAC,CAAA;QACxD,MAAM,CAAC,gBAAgB,CAAC,OAAO,EAAE,IAAI,CAAC,SAAS,EAAE,IAAI,CAAC,CAAA;IACxD,CAAC;IAED,qCAAoB,GAApB;QACE,MAAM,CAAC,mBAAmB,CAAC,OAAO,EAAE,IAAI,CAAC,WAAW,EAAE,IAAI,CAAC,CAAA;QAC3D,MAAM,CAAC,mBAAmB,CAAC,OAAO,EAAE,IAAI,CAAC,SAAS,EAAE,IAAI,CAAC,CAAA;IAC3D,CAAC;IAED,gCAAe,GAAf;QACE,EAAE,CAAC,CAAC,CAAC,IAAI,CAAC,KAAK,CAAC,KAAK,CAAC,CAAC,CAAC;YACtB,MAAM,CAAC,OAAO,IAAI,CAAC,KAAK,CAAC,WAAW,KAAK,QAAQ,CAAC,CAAC,CAAC,IAAI,CAAC,KAAK,CAAC,WAAW,CAAC,CAAC,CAAC,EAAE,CAAA;QACjF,CAAC;QAED,EAAE,CAAC,CAAC,CAAC,KAAK,CAAC,OAAO,CAAC,IAAI,CAAC,KAAK,CAAC,KAAK,CAAC,CAAC,CAAC,CAAC;YACrC,MAAM,CAAC,IAAI,CAAC,KAAK,CAAC,KAAK,CAAC,KAAK,CAAA;QAC/B,CAAC;QAED,MAAM,CAAK,IAAI,CAAC,KAAK,CAAC,KAAK,CAAC,GAAG,CAAC,UAAA,MAAM,IAAI,OAAA,MAAM,CAAC,KAAK,EAAZ,CAAY,CAAC,SAAE,IAAI,CAAC,IAAI,CAAC,CAAA;IACrE,CAAC;IAED,6BAAY,GAAZ,UAAa,MAAc;QACzB,EAAE,CAAC,CAAC,CAAC,KAAK,CAAC,OAAO,CAAC,IAAI,CAAC,KAAK,CAAC,KAAK,CAAC,CAAC,CAAC,CAAC;YACrC,IAAI,CAAC,KAAK,CAAC,QAAQ,CAAC,MAAM,CAAC,CAAA;YAC3B,MAAM,CAAA;QACR,CAAC;QAED,IAAM,WAAW,GAAW,IAAI,CAAC,KAAK,CAAC,KAAK,CAAC,OAAO,CAAC,MAAM,CAAC,CAAA;QAE5D,EAAE,CAAC,CAAC,WAAW,GAAG,CAAC,CAAC,CAAC,CAAC;YACpB,IAAI,CAAC,KAAK,CAAC,QAAQ,CAAK,IAAI,CAAC,KAAK,CAAC,KAAK,SAAE,MAAM,GAAE,MAAM,CAAC,UAAA,IAAI,IAAI,OAAA,CAAC,IAAI,CAAC,WAAW,EAAjB,CAAiB,CAAC,CAAC,CAAA;QACtF,CAAC;QAAC,IAAI,CAAC,CAAC;YACN,IAAI,CAAC,KAAK,CAAC,QAAQ,CAAK,IAAI,CAAC,KAAK,CAAC,KAAK,CAAC,KAAK,CAAC,CAAC,EAAE,WAAW,CAAC,QAAK,IAAI,CAAC,KAAK,CAAC,KAAK,CAAC,KAAK,CAAC,WAAW,GAAG,CAAC,CAAC,EAAE,CAAA;QAC9G,CAAC;IACH,CAAC;IAED,iCAAgB,GAAhB,UAAiB,MAAc;QAC7B,EAAE,CAAC,CAAC,CAAC,KAAK,CAAC,OAAO,CAAC,IAAI,CAAC,KAAK,CAAC,KAAK,CAAC,CAAC,CAAC,CAAC;YACrC,MAAM,CAAC,IAAI,CAAC,KAAK,CAAC,KAAK,KAAK,MAAM,CAAA;QACpC,CAAC;QAED,MAAM,CAAC,IAAI,CAAC,KAAK,CAAC,KAAK,CAAC,OAAO,CAAC,MAAM,CAAC,GAAG,CAAC,CAAC,CAAA;IAC9C,CAAC;IAEK,6BAAY,GAAlB,UAAmB,KAA6C;;;;;;wBAC9D,KAAK,CAAC,OAAO,EAAE,CAAA;wBAEf,EAAE,CAAC,CAAC,CAAC,CAAC,KAAK,CAAC,MAAM,YAAY,gBAAgB,CAAC,CAAC,CAAC,CAAC;4BAChD,MAAM,IAAI,KAAK,CAAC,8FAA8F,CAAC,CAAA;wBACjH,CAAC;6BAEG,IAAI,CAAC,KAAK,CAAC,QAAQ,EAAnB,wBAAmB;wBACrB,IAAI,CAAC,QAAQ,CAAC,cAAM,OAAA,CAAC,EAAE,QAAQ,EAAE,IAAI,EAAE,CAAC,EAApB,CAAoB,CAAC,CAAA;wBACzC,qBAAM,IAAI,CAAC,KAAK,CAAC,QAAQ,EAAE,EAAA;;wBAA3B,SAA2B,CAAA;;;wBAGvB,MAAM,GAAG,IAAI,MAAM,CAAC,KAAK,CAAC,MAAM,CAAC,KAAK,EAAE,GAAG,CAAC,CAAA;wBAClD,IAAI,CAAC,QAAQ,CAAC,cAAM,OAAA,CAAC,EAAE,MAAM,QAAA,EAAE,QAAQ,EAAE,KAAK,EAAE,CAAC,EAA7B,CAA6B,CAAC,CAAA;;;;;KACnD;IAEK,uBAAM,GAAZ;;;;;6BACM,CAAA,CAAC,IAAI,CAAC,KAAK,CAAC,IAAI,IAAI,IAAI,CAAC,KAAK,CAAC,OAAO,CAAA,EAAtC,wBAAsC;wBACxC,IAAI,CAAC,QAAQ,CAAC,cAAM,OAAA,CAAC,EAAE,QAAQ,EAAE,IAAI,EAAE,CAAC,EAApB,CAAoB,CAAC,CAAA;wBACzC,qBAAM,IAAI,CAAC,KAAK,CAAC,OAAO,EAAE,EAAA;;wBAA1B,SAA0B,CAAA;;;wBAE5B,IAAI,CAAC,QAAQ,CAAC,UAAA,SAAS,IAAI,OAAA,CAAC,EAAE,QAAQ,EAAE,KAAK,EAAE,IAAI,EAAE,CAAC,SAAS,CAAC,IAAI,EAAE,CAAC,EAA5C,CAA4C,CAAC,CAAA;;;;;KACzE;IAED,sBAAK,GAAL;QACE,IAAI,CAAC,QAAQ,CAAC,cAAM,OAAA,CAAC;YACnB,IAAI,EAAE,KAAK;SACZ,CAAC,EAFkB,CAElB,CAAC,CAAA;IACL,CAAC;IAED,uBAAM,GAAN;QAAA,iBAqCC;QApCC,MAAM,CAAC,CACL,oBAAC,wBAAS,IACR,QAAQ,EAAE,UAAA,SAAS,IAAI,OAAA,CAAC,KAAI,CAAC,SAAS,GAAG,SAAS,CAAC,EAA5B,CAA4B,EACnD,GAAG,EAAE,IAAI,CAAC,KAAK,CAAC,GAAG,EACnB,SAAS,EAAE,IAAI,CAAC,KAAK,CAAC,SAAS,EAC/B,QAAQ,EAAE,IAAI,CAAC,KAAK,CAAC,QAAQ,EAC7B,KAAK,EAAE,IAAI,CAAC,KAAK,CAAC,KAAK,EACvB,QAAQ,EAAE,IAAI,CAAC,KAAK,CAAC,QAAQ,EAC7B,IAAI,EAAC,SAAS,EACd,QAAQ,EAAE,CAAC,CAAC,EACZ,OAAO,EAAE,cAAM,OAAA,KAAI,CAAC,MAAM,EAAE,EAAb,CAAa;YAE5B,iCAAM,IAAI,CAAC,eAAe,EAAE,IAAI,IAAI,CAAC,KAAK,CAAC,WAAW,CAAO;YAC5D,IAAI,CAAC,KAAK,CAAC,OAAO,CAAC,MAAM,IAAI,IAAI,CAAC,KAAK,CAAC,IAAI,CAAC,CAAC,CAAC,CAC9C,oBAAC,sBAAO;gBACL,IAAI,CAAC,KAAK,CAAC,UAAU,IAAI,oBAAC,sBAAY,IAAC,QAAQ,EAAE,UAAC,CAAM,IAAK,OAAA,KAAI,CAAC,YAAY,CAAC,CAAC,CAAC,EAApB,CAAoB,GAAI;gBACtF,oBAAC,0BAAW,QACT,IAAI,CAAC,KAAK,CAAC,OAAO,CAAC,GAAG,CACrB,UAAC,MAAc;oBACb,OAAA,MAAM,CAAC,KAAK,CAAC,KAAK,CAAC,KAAI,CAAC,KAAK,CAAC,MAAM,CAAC,IAAI,CACvC,oBAAC,sBAAY,IACX,GAAG,EAAE,MAAM,CAAC,EAAE,EACd,OAAO,EAAE,cAAM,OAAA,KAAI,CAAC,YAAY,CAAC,MAAM,CAAC,EAAzB,CAAyB,EACxC,QAAQ,EAAE,KAAI,CAAC,gBAAgB,CAAC,MAAM,CAAC,IAEtC,MAAM,CAAC,KAAK,CACA,CAChB;gBARD,CAQC,CACJ,CACW,CACN,CACX,CAAC,CAAC,CAAC,CACF,EAAE,CACH,CACS,CACb,CAAA;IACH,CAAC;IACH,aAAC;AAAD,CAAC,AAlJD,CAAqB,KAAK,CAAC,SAAS,GAkJnC;AAED,kBAAe,MAAM,CAAA"}