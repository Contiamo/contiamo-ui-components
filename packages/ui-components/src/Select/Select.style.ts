import { hexOrColor, readableTextColor, spin, fadeIn, resetTransform } from "contiamo-ui-utils"

type Props = {
  theme: Theme
  color?: string
  disabled?: boolean
}

export default ({ theme, color, disabled }: Props): {} => {
  const backgroundColor = color && theme.colors ? hexOrColor(color)(theme.colors[color]) : "white"

  return {
    backgroundColor,
    position: "relative",
    display: "flex",
    alignItems: "center",
    padding: theme.spacing / 2,
    paddingRight: theme.spacing / 2 + 40,
    width: "fit-content",
    minHeight: 20,
    border: "1px solid",
    borderColor: theme.colors.grey30,
    opacity: disabled ? 0.5 : 1,
    cursor: "pointer",
    color: readableTextColor(backgroundColor)(["black", "white"]),
    outline: "none",
    pointerEvents: disabled ? "none" : "all",

    // downward caret.
    "&::after": {
      content: "''",
      position: "absolute",
      top: "50%",
      right: theme.spacing / 2,
      width: 0,
      height: 0,
      border: "4px solid transparent",
      borderTopColor: theme.colors.grey70,
      transform: "translateY(calc(-50% + 2px))"
    },

    // spinner when loading.
    "&.Select_updating::after": {
      top: 6,
      width: 16,
      height: 16,
      border: 0,
      borderRadius: "50%",
      boxShadow: `1px 0px 0px 0px ${theme.colors.grey70} inset`,
      animation: `.7s ${spin} linear infinite`
    },

    "& .Select__options": {
      position: "absolute",
      top: "calc(100% + 1px)",
      left: 0,
      zIndex: theme.baseZIndex + 100,
      width: "100%",
      boxShadow: "0 2px 7px 2px rgba(0, 0, 0, .14)",
      opacity: 0,
      transform: "translateY(-10px)",
      animation: `${fadeIn} .15s forwards ease,
        ${resetTransform} .15s forwards ease`
    },

    "& .Select__options_list": {
      maxHeight: "50vh"
    }
  }
}
