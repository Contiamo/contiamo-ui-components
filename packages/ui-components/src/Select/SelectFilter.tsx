import * as React from "react"
import glamorous, { GlamorousComponent } from "glamorous"
import { Theme } from "contiamo-ui-theme"

import { hexOrColor, darken } from "contiamo-ui-utils"

export interface IProps {
  css?: any
  key?: string | number
  className?: string
  placeholder?: string
  onChange?: (e: React.SyntheticEvent<HTMLInputElement>) => Promise<void>
  theme?: Theme
  color?: string
}

const Container = glamorous.div(({ theme, color }: { theme: Theme; color?: string }) => {
  const backgroundColor = color ? hexOrColor(color)(theme.colors.palette[color]) : "white"

  return {
    padding: 0,
    borderBottom: "1px solid",
    borderColor: darken(backgroundColor)(10),

    "& > input": {
      width: "100%",
      padding: theme.spacing / 2,
      border: 0,
      outline: "none",
      font: "inherit"
    }
  }
})

const SelectFilter: React.SFC<IProps> = ({ css, key, className, placeholder = "Filter...", onChange }: IProps) => (
  <Container key={key} css={css} className={className}>
    <input onClick={e => e.stopPropagation()} onChange={onChange} placeholder={placeholder} />
  </Container>
)

export default SelectFilter
