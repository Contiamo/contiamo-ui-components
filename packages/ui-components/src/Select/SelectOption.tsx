import * as React from "react"
import glamorous, { withTheme, GlamorousComponent } from "glamorous"
import { Theme } from "contiamo-ui-theme"

import { hexOrColor, readableTextColor, darken } from "contiamo-ui-utils"
import { Icon } from "contiamo-ui-components"

export interface IProps {
  key?: number | string
  css?: any
  className?: string
  selected?: boolean
  onClick?: () => void
  children?: React.ReactNode
  color?: string
}

export interface IPropsWithTheme extends IProps {
  theme: Theme
}

const Container = glamorous.div(
  ({ theme, color, selected }: { theme: Theme; color?: string; selected: boolean }): any => {
    const backgroundColor = color && theme.colors ? hexOrColor(color)(theme.colors.palette[color]) : "white"

    return {
      backgroundColor,
      position: "relative",
      padding: theme.spacing / 2,
      color: readableTextColor(backgroundColor)(["black", "white"]),
      outline: "none",

      ":hover": {
        backgroundColor: darken(backgroundColor)(5),
        color: readableTextColor(darken(backgroundColor)(5))(["black", "white"])
      },

      "&:not(:first-child)": {
        borderTop: "1px solid",
        borderColor: darken(backgroundColor)(10)
      }
    }
  }
)

const IconContainer = glamorous.div(({ theme }: { theme: Theme }): any => ({
  width: 8,
  height: 8,
  position: "absolute",
  top: "50%",
  right: 4,
  transform: "translate3d(-50%, -50%, 0)"
}))

const SelectOption: React.SFC<IPropsWithTheme> = ({
  css,
  key,
  className,
  theme,
  selected,
  color,
  onClick,
  children
}: IPropsWithTheme) => (
  <Container
    css={css}
    key={key}
    className={className}
    selected={!!selected}
    color={color}
    tabIndex={-2}
    role="option"
    aria-selected={selected}
    onClick={onClick}
  >
    {children}
    {selected ? (
      <IconContainer>
        <Icon name="X" size={theme.spacing} />
      </IconContainer>
    ) : null}
  </Container>
)

const WrappedSelectOption: React.SFC<IProps> = withTheme(SelectOption)

export default WrappedSelectOption
