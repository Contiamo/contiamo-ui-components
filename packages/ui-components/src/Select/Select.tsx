import * as React from "react"
import glamorous from "glamorous"

import SelectOption from "./Option/SelectOption"
import SelectFilter from "./Filter/SelectFilter"

import { Container, Options, OptionsList } from "./Select.style"

type Value = number | string

export interface Option {
  label?: string
  value?: Value
}

interface IProps {
  css?: any
  className?: string
  options: Option[]
  value: undefined | Value | Value[]
  filterable?: boolean
  disabled?: boolean
  onChange: (newValue: Value | Value[]) => void
  onClick?: () => void
  onFilter?: () => void
  color?: string
  placeholder?: string
}

interface IState {
  open: boolean
  updating: boolean
  filter: RegExp
}

class Select extends React.Component<IProps, IState> {
  state: IState = {
    open: false,
    updating: false,
    filter: new RegExp(/./)
  }

  // flow complains if this isn't initialized sooo...
  container: HTMLDivElement | null

  // This implements "click outside to close" behavior
  handleClick = (e: MouseEvent) => {
    const el = this.container

    // if we're somehow not working with a DOM node (flowtype is fun!)
    if (!(e.target instanceof Node)) {
      return
    }
    // if we're clicking on the Select itself,
    if (el && el.contains(e.target)) {
      return
    }

    // if we're clicking outside,
    this.close()
  }

  handleEsc = (e: KeyboardEvent) => {
    if (e.keyCode === 27) {
      this.close()
    }
  }

  componentDidMount() {
    window.addEventListener("click", this.handleClick, true)
    window.addEventListener("keyup", this.handleEsc, true)
  }

  componentWillUnmount() {
    window.removeEventListener("click", this.handleClick, true)
    window.removeEventListener("keyup", this.handleEsc, true)
  }

  getDisplayValue(): string {
    if (!this.props.value) {
      return this.props.placeholder || "No entries selected"
    }

    if (!Array.isArray(this.props.value)) {
      return this.props.options.filter(option => option.value === this.props.value)[0].label
    }

    const listDisplay = this.props.options
      .map(option => ((this.props.value as Value[]).indexOf(option.value) > -1 ? option.label : null))
      .filter(a => !!a)
      .join(", ")
    return listDisplay === "" ? this.props.placeholder || "No entries selected" : listDisplay
  }

  selectOption(option: Option) {
    if (!Array.isArray(this.props.value)) {
      this.props.onChange(option.value)
      return
    }

    const optionIndex: number = this.props.value.indexOf(option.value)

    if (optionIndex < 0) {
      this.props.onChange([...this.props.value, option.value])
    } else {
      this.props.onChange([...this.props.value.slice(0, optionIndex), ...this.props.value.slice(optionIndex + 1)])
    }
  }

  isOptionSelected(option: Option) {
    if (!Array.isArray(this.props.value)) {
      return this.props.value === option
    }

    return this.props.value.indexOf(option.value) > -1
  }

  async updateFilter(event: React.SyntheticEvent<HTMLInputElement>) {
    event.persist()

    if (!(event.target instanceof HTMLInputElement)) {
      throw new Error("<Select>: Your filter field is _not_ an input element and therefore has an unreadable value.")
    }

    if (this.props.onFilter) {
      this.setState(() => ({ updating: true }))
      await this.props.onFilter()
    }

    const filter = new RegExp(event.target.value, "i")
    this.setState(() => ({ filter, updating: false }))
  }

  async toggle() {
    if (!this.state.open && this.props.onClick) {
      this.setState(() => ({ updating: true }))
      await this.props.onClick()
    }
    this.setState(prevState => ({ updating: false, open: !prevState.open }))
  }

  close() {
    this.setState(() => ({
      open: false
    }))
  }

  render() {
    return (
      <Container
        innerRef={container => (this.container = container)}
        css={this.props.css}
        className={this.props.className}
        updating={this.state.updating}
        color={this.props.color}
        disabled={this.props.disabled}
        role="listbox"
        tabIndex={-2}
        onClick={() => this.toggle()}
      >
        <div>{this.getDisplayValue() || this.props.placeholder}</div>
        {this.props.options.length && this.state.open ? (
          <Options>
            {this.props.filterable && <SelectFilter onChange={(e: any) => this.updateFilter(e)} />}
            <OptionsList>
              {this.props.options.map(
                (option: Option) =>
                  option.label.match(this.state.filter) && (
                    <SelectOption
                      key={String(option.value)}
                      onClick={() => this.selectOption(option)}
                      selected={this.isOptionSelected(option)}
                    >
                      {option.label}
                    </SelectOption>
                  )
              )}
            </OptionsList>
          </Options>
        ) : (
          ""
        )}
      </Container>
    )
  }
}

export default Select
