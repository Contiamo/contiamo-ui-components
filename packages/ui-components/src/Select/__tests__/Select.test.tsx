import * as React from "react"
import { render } from "enzyme"

import { Select, Option } from "../Select"
import style from "../Select.style"

const options: [Option] = [
  { id: 1, label: "John", value: -10 },
  { id: 2, label: "Joey", value: "Nein" },
  { id: 3, label: "Tupac", value: true },
  { id: 4, label: "Chandler", value: [{ alive: true }] }
]

describe("Select", () => {
  it("Should render correctly", () => {
    expect(render(<Select options={options} disabled filterable multiple placeholder="Select me" />)).toMatchSnapshot()
  })
})
