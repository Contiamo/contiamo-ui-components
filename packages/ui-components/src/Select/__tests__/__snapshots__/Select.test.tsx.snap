// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Should render correctly 1`] = `
<div
  class="css-3kwy5e"
  role="listbox"
  tabindex="-2"
>
  <div>
    Select me
  </div>
</div>
`;
